File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/analytics/AnalyticsService.java
package com.pbshop.java.spring.maven.jpa.postgresql.analytics;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.product.PriceEntry;
import com.pbshop.java.spring.maven.jpa.postgresql.product.PriceEntryRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Product;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductRepository;

@Service
@Transactional
public class AnalyticsService {

    private final ProductRepository productRepository;
    private final PriceEntryRepository priceEntryRepository;

    public AnalyticsService(ProductRepository productRepository, PriceEntryRepository priceEntryRepository) {
        this.productRepository = productRepository;
        this.priceEntryRepository = priceEntryRepository;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> lowestEver(Long productId) {
        Product product = requireProduct(productId);
        PriceEntry lowest = priceEntryRepository.findByProductIdOrderByPriceAsc(productId).stream()
                .min(Comparator.comparing(PriceEntry::getPrice))
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "가격 이력을 찾을 수 없습니다."));

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("productId", product.getId());
        response.put("productName", product.getName());
        response.put("lowestPrice", lowest.getPrice().setScale(2, RoundingMode.HALF_UP));
        response.put("sellerId", lowest.getSeller().getId());
        response.put("checkedAt", lowest.getCheckedAt() == null ? null : lowest.getCheckedAt().toString());
        return response;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> unitPrice(Long productId) {
        Product product = requireProduct(productId);
        BigDecimal price = priceEntryRepository.findByProductIdOrderByPriceAsc(productId).stream()
                .map(PriceEntry::getPrice)
                .min(Comparator.naturalOrder())
                .orElse(BigDecimal.ZERO)
                .setScale(2, RoundingMode.HALF_UP);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("productId", product.getId());
        response.put("productName", product.getName());
        response.put("unit", "ea");
        response.put("quantity", 1);
        response.put("unitPrice", price);
        return response;
    }

    private Product requireProduct(Long productId) {
        return productRepository.findById(productId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "상품을 찾을 수 없습니다."));
    }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/analytics/AnalyticsService.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.analytics

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.PbShopException
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.StubResponse
import io.ktor.http.HttpStatusCode

class AnalyticsService(
    private val repository: AnalyticsRepository,
) {
    fun lowestEver(productId: Int): StubResponse {
        val record =
            repository.findLowestEver(productId)
                ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "상품 가격 분석 정보를 찾을 수 없습니다.")
        return StubResponse(
            data = mapOf(
                "isLowestEver" to (record.currentPrice <= record.lowestPrice),
                "currentPrice" to record.currentPrice,
                "lowestPrice" to record.lowestPrice,
                "lowestDate" to record.lowestDate?.toString(),
            )
        )
    }

    fun unitPrice(productId: Int): StubResponse {
        val record =
            repository.findUnitPrice(productId)
                ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "단가 계산 정보를 찾을 수 없습니다.")
        return StubResponse(
            data = mapOf(
                "unitPrice" to record.unitPrice,
                "unit" to record.unit,
                "quantity" to record.quantity,
            )
        )
    }
}
